### PR TITLE
preserve kwargs order on assign func for py36plus - #14207

### LINKED ISSUE #14207: API: allow dependent assignment?
Since kwarg order will be [guaranteed](https://docs.python.org/3.6/whatsnew/3.6.html#pep-468-preserving-keyword-argument-order) in python 3.6, we could allow this...though would only want to try with a 3.6+ version check, otherwise code could more or less randomly work/not work, depending on dict order.

``` python
In [44]: df = pd.DataFrame({'a': [1,2,3]})

In [45]: df.assign(b=1, c=lambda x: x['b'] * 2)
```

### PR DESCRIPTION
**Pull request by bobhaffner**
- [x] closes #14207
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v0.21.0.txt [hunk 1]
```diff
@@ -117,6 +117,7 @@ Other Enhancements
 - :func:`MultiIndex.is_monotonic_decreasing` has been implemented.  Previously returned ``False`` in all cases. (:issue:`16554`)
 - :func:`Categorical.rename_categories` now accepts a dict-like argument as `new_categories` and only updates the categories found in that dict. (:issue:`17336`)
 - :func:`read_excel` raises ``ImportError`` with a better message if ``xlrd`` is not installed. (:issue:`17613`)
+- :func:`assign` will preserve the original order of **kwargs for Python 3.6+ users
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.21.0.txt]: This will be

    :meth:`DataFrame.assign`

And put `**kwargs` in double backticks. And maybe reference the original behavior like "instead of sorting the column names".

File: pandas/core/frame.py [hunk 2]
```diff
@@ -2576,10 +2576,10 @@ def assign(self, **kwargs):
...
         Since ``kwargs`` is a dictionary, the order of your
-        arguments may not be preserved. To make things predicatable,
-        the columns are inserted in alphabetical order, at the end of
-        your DataFrame. Assigning multiple columns within the same
-        ``assign`` is possible, but you cannot reference other columns
+        arguments may not be preserved if you are using Python 3.5 and earlier.
+        To make things predicatable, the columns are inserted in alphabetical
+        order, at the end of your DataFrame. Assigning multiple columns within
+        the same ``assign`` is possible, but you cannot reference other columns
         created within the same ``assign`` call.
```
Comment (TomAugspurger) [inline pandas/core/frame.py]: I would rephrase the first couple sentences to say something like

"For python 3.6 and above, the columns are inserted in the order of **kwargs. For python 3.5 and earlier, since **kwargs is unordered, the columns are inserted in alphabetical order at the end of your DataFrame.`"

And then pick up with "Assigning multiple..."

That emphasizes that 3.6+ is good :)

**Review by TomAugspurger — APPROVED** [review]

File: pandas/core/frame.py [hunk 3]
```diff
@@ -2620,14 +2620,18 @@ def assign(self, **kwargs):
         data = self.copy()
 
         # do all calculations first...
-        results = {}
+        results = OrderedDict()
         for k, v in kwargs.items():
             results[k] = com._apply_if_callable(v, data)
 
+        # sort by key for 3.5 and earlier, but preserve order for 3.6 and later
+        if sys.version_info <= (3, 5):
```
Comment (jreback) [inline pandas/core/frame.py]: use ``pandas.compat.PY36`` here

File: pandas/tests/frame/test_mutate_columns.py [hunk 4]
```diff
@@ -61,14 +62,23 @@ def test_assign_multiple(self):
...
 
-    def test_assign_alphabetical(self):
+    def test_assign_order(self):
         # GH 9818
         df = DataFrame([[1, 2], [3, 4]], columns=['A', 'B'])
         result = df.assign(D=df.A + df.B, C=df.A - df.B)
-        expected = DataFrame([[1, 2, -1, 3], [3, 4, -1, 7]],
-                             columns=list('ABCD'))
+
+        if sys.version_info <= (3, 5):
```
Comment (jreback) [inline pandas/tests/frame/test_mutate_columns.py]: see above

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by TomAugspurger** [comment]
This should apply to py36+, and we don't want to have to update this with
every python release. We *could* add a `PY_LT_36` in pandas.compat`, but
that seems like overkill for now.

**Comment by jreback** [comment]
> This should apply to py36+, and we don't want to have to update this with
every python release. We *could* add a `PY_LT_36` in pandas.compat`, but
that seems like overkill for now.

we already have this which is ``PY36`` means >= 3.6, or you can do ``not PY35`` which is < 3.6.

**Comment by TomAugspurger** [comment]
Ah, I thought those were exact. Yep let's do that then.

**Comment by bobhaffner** [comment]
Thanks for the feedback, gents.  I'll get these fixed

**Review by jorisvandenbossche — APPROVED** [review]

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @bobhaffner

**Comment by bobhaffner** [comment]
It's the least I could do, @jreback.  Thanks to you and @TomAugspurger for the guidance and everything else you do for the project
